### PR TITLE
Add connections to cloud storage providers to browser

### LINKED ISSUE #35450: Add S3 to Browser panel
**Feature description.**
The Browser panel is neat, I can list a local directory, list all my PostGIS databases and tables, etc. it would be nice to have an S3 item here where you could add buckets, then expanding each bucket would list all the files in the bucket just like the file directory.

This makes it a lot easier and faster to work with data stored in S3 buckets without needing to manually enter the bucket name and object key for each file in the add vector/raster dialog.

A workaround is to mount the bucket on your local file system using something like https://github.com/s3fs-fuse/s3fs-fuse, but that takes configuration and setup, so something out of the box with QGIS would be nice.

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR introduces support for storing and browsing cloud storage connections to the QGIS browser panel. GDAL supported vector and raster data sources can be loaded directly from the cloud provider (we rely on GDAL's VSI cloud handlers for this functionality). Supports all GDAL VSI cloud handlers, including S3, Azure, Google Drive, etc. 

Connections can be created, modified, removed, and exported/imported to/from XML files. 

https://github.com/qgis/QGIS/assets/1829991/8f181bda-697e-4c13-b8e0-ac16cf933011

https://github.com/qgis/QGIS/assets/1829991/ee39e34f-534b-4a1e-8d22-b96577a1739c

Fixes https://github.com/qgis/QGIS/issues/35450

### CODE REVIEW

**Comment by uclaros** [comment]
Nice!!
Are copc files handled as well, or is it gdal vsi supported formats only?

**Comment by nyalldawson** [comment]
@uclaros only gdal/ogr sources -- we're relying on gdal's vsi driver here

**Review by elpaso — COMMENTED** [review]
LGTM, can you please attach a snapshot of the new GUI ?

**Comment by nyalldawson** [comment]
@elpaso 

> can you please attach a snapshot of the new GUI ?

I've added a screencast to the PR comment

File: src/core/providers/gdal/qgsgdalcloudconnection.cpp [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+  QgsGdalCloudProviderConnection::Data conn;
+  conn.vsiHandler = dsUri.param( QStringLiteral( "handler" ) );
+  conn.container = dsUri.param( QStringLiteral( "container" ) );
+  conn.rootPath = dsUri.param( QStringLiteral( "rootPath" ) );
+
+  const QStringList credentialOptions = dsUri.param( QStringLiteral( "credentialOptions" ) ).split( '|' );
+  for ( const QString &option : credentialOptions )
+  {
+    const thread_local QRegularExpression credentialOptionKeyValueRegex( QStringLiteral( "(.*?)=(.*)" ) );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/providers/gdal/qgsgdalcloudconnection.cpp:73]: Given it's thread_local there's no performance benefit to that, so I'd prefer to place it as close to the variable use as possible

**Comment by Youssef-Harby** [comment]
In the AWS option, how can I set AWS_S3_ENDPOINT according to [GDAL Docs](https://gdal.org/en/stable/user/virtual_file_systems.html#vsis3-aws-s3-files)
I want to connect Hetzner, Contabo or Minio bucket, that's the reason I am asking

**Comment by agiudiceandrea** [comment]
> Nice!! Are copc files handled as well, or is it gdal vsi supported formats only?

See https://github.com/qgis/QGIS/issues/62644.

**Pull request merged**